interfaces/builtin/utils.go

```diff
@@ -125,3 +126,49 @@ func verifySlotPathAttribute(slotRef *interfaces.SlotRef, attrs interfaces.Attre
 	}
 	return cleanPath, nil
 }
+
+// aareExclusivePatterns takes a string and generates deny alternations. Eg,
+// aareExclusivePatterns("foo") returns:
+// []string{
+//   "[^f]*",
+//   "f[^o]*",
+//   "fo[^o]*",
+// }
+func aareExclusivePatterns(orig string) []string {
+	s := make([]string, len(orig))
+	prefix := ""
+	for i, letter := range orig {
+		if i > 0 {
+			prefix = orig[:i]
+		}
+		s[i] = fmt.Sprintf("%s[^%c]*", prefix, letter)
+	}
+	return s
+}
+
+// getDesktopFileRules(<snap instance name>) generates snippet rules for
+// allowing access to the specified snap's desktop files in
+// dirs.SnapDesktopFilesDir, but explicitly denies access to all other snaps'
+// desktop files since xdg libraries may try to read all the desktop files
+// in the dir, causing excessive noise. (LP: #1868051)
+func getDesktopFileRules(snapInstanceName string) []string {
+	baseDir := dirs.SnapDesktopFilesDir
+
+	rules := []string{
+		"# Support applications which use the unity messaging menu, xdg-mime, etc",
+		"# This leaks the names of snaps with desktop files",
+		fmt.Sprintf("%s/ r,", baseDir),
+		"# Allowing reading only our desktop files (required by (at least) the unity",
+		"# messaging menu).",
+		"# parallel-installs: this leaks read access to desktop files owned by keyed",
+		"# instances of @{SNAP_NAME} to @{SNAP_NAME} snap",
+		fmt.Sprintf("%s/@{SNAP_INSTANCE_NAME}_*.desktop r,", baseDir),
+		"# Explicitly deny access to other snap's desktop files",
+		fmt.Sprintf("deny %s/@{SNAP_INSTANCE_NAME}[^_.]*.desktop r,", baseDir),
```

Comment:
I forget how apparmor regex's work, does this match any file ending in `.desktop` i.e. /var/lib/snapd/applications/foobar123Zdesktop`, where the `.` before `desktop` matches any character? or does apparmor regex treat `.` as verbatim only the `.` character ?

Reply:
AppArmor does not use regexps: the syntax is closer (but not identical) to glob patterns.  So the dots are just dots.

Reply:
ah ok, thanks for clarifying this point